### PR TITLE
ARM wheels

### PR DESCRIPTION
**Pull request by anivegesana**
This PR builds the wheels for the ARM architecture and makes some minor rearrangements in the README to bring the parts that talk about Python into the same section. It adds a autogenerated `_version.py` file which will contain the version of the wheel, which can be automatically updated from the Git commit hash or tag.

I did not compile the Windows ARM wheels because numpy doesn't ship Windows ARM wheels. The Linux ARM wheels are only compiled on master to save build time. A single Linux ARM build is 20 minutes. The only ARM build that is run routinely is the Mac one, since M1/M2 Macs are the latest generation of Macs.

### CODE REVIEW

**Comment by anivegesana** [comment]
I think Mac still takes too long. I think it is best to drop Mac x86 from non-master branch builds. I do not want to rebuild this, but here are the changes that can be made to make it save time.

<details>
<summary>Drop Mac x86 on debug builds</summary>

```patch
diff --git a/.github/workflows/build_wheels.yml b/.github/workflows/build_wheels.yml
index e476f70..6798886 100644
--- a/.github/workflows/build_wheels.yml
+++ b/.github/workflows/build_wheels.yml
@@ -52,13 +52,16 @@ jobs:
           platforms: all
 
       - name: Build wheels (development)
-        if: github.ref != 'refs/heads/master'
+        if: github.ref != 'refs/heads/master' && !startsWith(github.ref, 'refs/tags/v')
         uses: pypa/cibuildwheel@v2.11.2
+        env:
+          CIBW_ARCHS_MACOS: "arm64"
 
       - name: Build wheels (production)
-        if: github.ref == 'refs/heads/master'
+        if: github.ref == 'refs/heads/master' || startsWith(github.ref, 'refs/tags/v')
         uses: pypa/cibuildwheel@v2.11.2
         env:
+          CIBW_ARCHS_MACOS: "x86_64 arm64"
           CIBW_ARCHS_LINUX: "auto aarch64"
 
       - uses: actions/upload-artifact@v3
diff --git a/pyproject.toml b/pyproject.toml
index 0024880..b82ca36 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -53,8 +53,10 @@ write_to = "pythonmodule/_version.py"
 build = "cp36-* cp38-macosx_arm64" # pp39-* cp39-win_arm64
 skip = "*-win32 *_i686 *-musllinux_*"
 
-[tool.cibuildwheel.macos]
-archs = ["x86_64", "arm64"]
+# We will only build x86_64 when merged into master to save on compilation
+# time.
+# [tool.cibuildwheel.macos]
+# archs = ["x86_64", "arm64"]
 
 # We will only build aarch64 when merged into master to save on compilation
 # time.
```

</details>

What do you think is better? Having a separate branch for staging the more time-intensive builds or working mainly on other branches and merging into master when it is working on the more easy to compile subset (4 minutes per build amounting to 1 hour of free credits)?

**Comment by wangyiqiu** [comment]
Hi @anivegesana good suggestion.

Yes, I agree that let's work mainly on other branches, and put the time-insensitive builds on the master. You could change the destination of this PR to the master branch.

**Comment by anivegesana** [comment]
Hey @wangyiqiu, I create a PyPI username. It matches my Github username: @anivegesana. I would be happy to be a maintainer on dbscan, but I don't know how much maintenance this project will require. Using the Python limited ABI and the CI reduce the amount of effort required to publish to PyPI because the wheels are built automatically and, unlike Cython extensions, are portable across Python versions, so we do not need to periodically recompile/update the package to allow users of new Python versions to use it. We compile dbscan once and it is ready forever.

The only real maintenance would be adding features or bug fixes. I will read the paper and get accustomed to the dbscan folder itself. (I have only looked at the outward facing Python package up until now.) I will also follow the issues and open pull requests to try to resolve them.

Thank you!

**Comment by wangyiqiu** [comment]
Great thanks @anivegesana  -- I have sent you an invite on PyPI.

As you may have already know, the code stems from one of the methods in this [paper](https://arxiv.org/abs/1912.06255), and the goal of this repository is to enable more people to use the code in real applications. I really like your work in making the code more accessible by modifying the API, streamlining the build, improving the portability and compatibility. They are very important for making a bigger impact.

I think the short-term goal is to make v0.0.10 work well on most platforms / environment without bugs like you have mentioned. Looking forward, it may also be interesting to support dimensional beyond 20 from a software standpoint. Potential solutions may include falling back to other existing DBSCAN implementations when the dimensionality is high, or write our own fast DBSCAN code for higher dimensional data.

**Comment by anivegesana** [comment]
@wangyiqiu Maybe it would be a good idea for us to look into [manually triggering](https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment#creating-an-environment) builds on master. There is a rather [limited budget](https://docs.github.com/en/billing/managing-billing-for-github-actions/about-billing-for-github-actions#included-storage-and-minutes) for number of free minutes and it would be irritating to not be able to build things if the minutes run out.

**Comment by wangyiqiu** [comment]
That's a good idea, though I did not figure out how to do that. At the moment I will upgrade to Pro if the minutes run out.

**Comment by anivegesana** [comment]
I can also look into other CIs like Circle CI and Cirrus CI. Worst comes to worst, we can just disable MacOS builds on master and development and I could periodically build them on my MacBook. These cost the most because Apple makes it hard to reverse-engineer their products and create virtual machines, so the builds are being done on physical Mac devices. This additional cost would be the same across all CIs. The only difference would be how much free credit they give a month. This would be inconvenient though, but it would work in the short term.

**Comment by anivegesana** [comment]
@wangyiqiu Considering that we don't seem to use that many builds a month, I don't think it will be a big concern for now. I can look into it a bit later. You can go ahead and push the release.

**Comment by wangyiqiu** [comment]
Sounds good, I agree.

I just pushed a branch called 0.0.11-dev. I thought of a few items that we can include in this update (including the one you mentioned):

1. Support float32, int32 and/or int64 in the Python API where the basic idea is to cast non-double input into doubles. I tried to work with [this function call in dbscanmodule.cpp](https://github.com/wangyiqiu/dbscan-python/blob/a0951f22b712b793919268719401cb97fa155249/src/dbscanmodule.cpp#L22) but did ont get it to work. I tried to instead call a `PyArray_CheckFromAny` function followed by some type casts, but the data seems to be garbled at the beginning. Are you familiar with how this works? Also, is it easier if we handle the data type conversion in Python rather than C++?
3. Update example.py like you have mentioned. Let's test most / all the wheels before publishing 0.0.11.
4. Update running time diagram in case the performance properties changed since 0.0.10.
5. Use our current README as a description for the PyPI repository.

Any thoughts?

**Comment by anivegesana** [comment]
Hi Yiqiu, thanks for the update! (pun intended)

1. I could be wrong, but `PyArray_FROMANY` creates a new array object if the dtype doesn't match. It isn't explicitly stated in the C or Python documentation for that function, but there is an example for [`np.asarray`](https://numpy.org/doc/stable/reference/generated/numpy.asarray.html).
```py
import numpy as np
from sklearn.datasets import make_blobs
from sklearn.preprocessing import StandardScaler
import dbscan

centers = [[1, 1], [-1, -1], [1, -1]]
X, labels_true = make_blobs(n_samples=750, centers=centers,
                            cluster_std=0.4, random_state=0)
X = StandardScaler().fit_transform(X)

X_labels, X_core = dbscan.DBSCAN(X, eps=0.3, min_samples=10)

Y = (X * 100000).astype(np.int64)
Y_labels, Y_core = dbscan.DBSCAN(Y, eps=30000, min_samples=10)
assert (X_labels == Y_labels).all() and (X_core == Y_core).all()
```
2.  Well, I believe that `example.py` is still useful for people to be able to understand how to use the library. I think we should have a separate file `test_python_module.py` that contains an almost exact copy of `example.py`, but can be run without matplotlib.
3. Yep. It is a good idea to retime everything and make sure that the figure is accurate and truthful and it didn't get slower. I think if it did get slower, however, it is not likely due to significant overhead from the limited API because it should only add a couple of additional function calls which would be many times smaller than the number of operations in the DBSCAN code itself. If there is a big timing difference, we should investigate to see if we added a bug in between versions.
4. 👍

**Comment by anivegesana** [comment]
Fixed the problems mentioned above (except for the CPU usage bug.) The Python test case are run during the build phase of cibuildwheel. I manually verified that the METADATA of the wheel does have the README information.

Here is the build: https://github.com/anivegesana/dbscan-python/actions/runs/3671346437/jobs/6206487165

**Comment by wangyiqiu** [comment]
Thanks! @anivegesana I'm still looking into it and will get back to you asap.

**Comment by anivegesana** [comment]
The other option is to drop older platforms and upgrade to C++20. Right now, I use defaults for a lot of the cibuildwheel settings. For most of the platforms, the wheels are compiled on Python 3.6. Changing the settings to compile on 3.8 (so that the versions of GCC match on Linux) wouldn't change anything for the end user since we never promised supporting older versions than that. The problem is with dropping older OS versions (like MacOS Catalina and Linuxes with GCC<12.) We could do it, but I do not know if people are using outdated OSes to install this package and it might be a good idea to leave this small buffer for a year or two in case they are.

**Comment by anivegesana** [comment]
I think it might be possible to just stick with C++17 and still remove the busy waiting in the scheduler. This pattern seems to do the same thing as the C++20 `atomic_wait`, but I have only tried it out on macOS Big Sur.
```c++
#ifdef __cpp_lib_atomic_wait
  std::atomic<int> deque_size;
#else
  struct : std::atomic<int> {
    std::condition_variable_any cv{};

    inline void notify_one() {
      cv.notify_one();
    }
    inline void notify_all() {
      cv.notify_all();
    }
    inline void wait(int val) {
      static struct {
        void lock() {}
        void unlock() {}
      } mutex{}; // a dummy mutex

      bool waited = false;
      if (load() == val)
        cv.wait(mutex, [&]() {
          bool old_waited = waited;
          waited = true;
          return old_waited;
        });
    }
  } deque_size;
#endif
```
I am trying to see if I can verify that this performs the same operation as `atomic_wait` and if so, that would save the headache of vendoring the `atomic_wait` backport. I also am now keeping track of the number of jobs available because that seems to be working better. Still do not know if this is equivalent/works. Will look into verifying that.

**Comment by wangyiqiu** [comment]
Hi @anivegesana thanks for finding this issue!
I tested the python module and indeed, the scheduler does not stop after the DBSCAN computation.

The parlay/pbbs scheduler is designed for C++ programs, such that a unique copy of the scheduler is created globally, and the scheduler is only destroyed at program exit. The problem you identified points to the high utilization of the CPU between the end of the DBSCAN function call and the end of the Python session, and this is because the scheduler is still expecting more work (and the threads are spinning).

The solution you proposed is a good one in the sense that it still keeps the scheduler active in the background, but potentially without causing high CPU-utilization before the next DBSCAN call. The downside, like you said, being that it may require some comprehensive testing of the scheduler, which can be quite a lot of work.

I proposed a quick fix (https://github.com/wangyiqiu/dbscan-python/commit/9a27ef4b8f04e983a24138ab17748be2f8df443e) by adding a start and stop function for the scheduler, and calling them at the correct places. Essentially, a new copy of the scheduler is responsible for each DBSCAN function call, rather than one scheduler being active during the entire Python session. The downside is that we need to start and stop the scheduler at the correct places in the code, and maybe there is some scheduler initialization overhead.

For the work (using conditional variable and atomic wait) you have in progress, I think it is great and we can keep it for future releases. I also encourage you to submit a PR for the [parlaylib](https://github.com/cmuparlay/parlaylib) to help them fix the same scheduler issue. Let me know what you think.

**Comment by anivegesana** [comment]
Happy new year, @wangyiqiu!

Sounds good. It would be easier to just add a temporary fix and create an issue on the parlaylib repo. Eventually, when they get a general solution, we can upgrade dbscan to use the latest parlaylib version.

**Comment by wangyiqiu** [comment]
Sounds good @anivegesana Happy new year to you, too!

**Comment by anivegesana** [comment]
Created the issue at https://github.com/cmuparlay/parlaylib/issues/33. I will test my solution against their test cases and create a PR there after I get it to pass. I think the workaround 9a27ef4b8f04e983a24138ab17748be2f8df443e that you pushed yesterday is good enough for this package, if you want to push another release.

**Comment by wangyiqiu** [comment]
Thanks, it looks good.

I published v0.0.11, and it seems to work well. Before publishing it I encountered some errors which seems to be coming from Python test configuration: https://github.com/wangyiqiu/dbscan-python/actions/runs/3858728146
Currently I temporarily disabled the Python tests. Will look into it soon. 

I also created a v0.0.12 branch for the next version: https://github.com/wangyiqiu/dbscan-python/tree/0.0.12-dev

**Comment by anivegesana** [comment]
Hmm. That looks odd. I guess that the Python tests weren't being run on non-master branches so I didn't notice that I removed the "py36" extra that the test cases use.

`CIBW_TEST_EXTRAS: "test36"`

The extras are specified in setup.py. I was supposed to add another extra `'py36': ['scikit-learn', 'matplotlib', 'pytest'],`.

**Comment by anivegesana** [comment]
Somewhat related to the issue from yesterday, it might be a good idea to swap Option 1 and Option 2 in the README. Putting the C++ instructions first might make it seem like that there is some benefit in rebuilding the C++ version over the prebuilt Python one. Let me know if you think this is a good idea.

**Comment by wangyiqiu** [comment]
Hi @anivegesana 
Thanks for the suggestions!

Yes let's swap option 1 and 2. I agree we should let the users use the prebuilt Python. Like you suggested, let's also swap the current description with `pip install -e .`.

I had actually being testing against the sequential sklearn since I did not know they can also run in parallel! I will rerun the tests to compare against the parallel sklearn and update the figure / number.

**Comment by wangyiqiu** [comment]
Sounds good, I will make four columns. Some of the bars can be quite similar. I will use the log scale when that happens.
